File: sorting/shellsort.c
#include "utility.h"
int spseq[8] = {701, 301, 132, 57, 23, 10, 4, 1};
void shellsort(int *arr, int lo, int hi) {
	//for getting gaps
	for (int x = 0; x <  8; x++) {
		//takes up element group for a gap
		for (int offset = lo; offset < spseq[x]; offset++) {
			//for jumping between element groups
			for (int i = offset + spseq[x]; i <= hi; i+= spseq[x]) {
				int j = i - spseq[x];
				int temp = arr[i];
				for (; j >= lo && arr[j] > temp; j -= spseq[x]) {
					arr[j + spseq[x]] = arr[j];
				}
				arr[j + spseq[x]] = temp;
			}
		}
	}
}
int main() {
	srand(time(0));
	int upper = 100, lower = 0;
	int arr[MAX];
	for (int i = 0; i < MAX; i++) {
		arr[i] = (rand() % (upper - lower + 1)) + lower;
	}

	printarray(arr, MAX);
	shellsort(arr, 0, MAX - 1);
	printarray(arr, MAX);
}

File: sorting/shellsort.js
let gaps = [ 701, 301, 132, 57, 23, 10, 4, 1 ];
let arr = [];
const max = 10;
for (let i = 0; i < max; i++) {
	arr.push(Math.floor(Math.random() * 100));
}
console.log('before');
console.log(arr);
for (let gap of gaps) {
	for (let offset = 0; offset < gap; offset++) {
		for (let j = offset; j < max; j += gap) {
			let k = j + gap;
			let temp = arr[k];
			for (; k >= gap && arr[k - gap] > temp; k -= gap) {
				arr[k] = arr[k - gap];
			}
			arr[k] = temp;
		}
	}
}

function sorted(arr, max) {
	for (let i = 1; i < max; i++ ) {
		if (arr[i] < arr[i - 1])
			return false
	}
	return true;
}
console.log(arr);
console.log(sorted(arr, max));
